algorithms/kernel/dtrees/forest/oneapi/cl_kernels/df_common_kernels.cl

```diff
@@ -0,0 +1,227 @@
+
+    __kernel void radixScan(const __global radixIntType * values, __global int * partialHists, int nRows, int bitOffset) {
+        const int RADIX_BITS = 4;
+
+        const int n_groups             = get_num_groups(0);
+        const int n_sub_groups         = get_num_sub_groups();
+        const int n_total_sub_groups   = n_sub_groups * n_groups;
+        const int nElementsForSubgroup = nRows / n_total_sub_groups + !!(nRows % n_total_sub_groups);
+        const int local_size           = get_sub_group_size();
+
+        const int id           = get_local_id(0);
+        const int local_id     = get_sub_group_local_id();
+        const int sub_group_id = get_sub_group_id();
+        const int group_id     = get_group_id(0) * n_sub_groups + sub_group_id;
+
+        int iStart = group_id * nElementsForSubgroup;
+        int iEnd   = (group_id + 1) * nElementsForSubgroup;
+
+        if (iEnd > nRows)
+        {
+            iEnd = nRows;
+        }
+
+        int offset[1 << RADIX_BITS];
+        const int radix_range   = 1 << RADIX_BITS;
+        const int radix_range_1 = radix_range - 1;
+        for (int i = 0; i < radix_range; i++)
+        {
+            offset[i] = 0;
+        }
+
+        for (int i = iStart + local_id; i < iEnd; i += local_size)
+        {
+            radixIntType data_bits = ((invBits(values[i]) >> bitOffset) & radix_range_1);
+            for (int j = 0; j < radix_range; j++)
+            {
+                int value          = data_bits == j;
+                int partial_offset = sub_group_reduce_add(value);
+                offset[j] += partial_offset;
+            }
+        }
+
+        if (local_id == 0)
+        {
+            for (int j = 0; j < radix_range; j++)
+            {
+                partialHists[group_id * radix_range + j] = offset[j];
+            }
+        }
+    }
+
+    __kernel void radixHistScan(const __global int * partialHists, __global int * partialPrefixHists, int nSubgroupSums) {
+        const int RADIX_BITS = 4;
+
+        if (get_sub_group_id() > 0) return;
+
+        const int local_size = get_sub_group_size();
+        const int local_id   = get_sub_group_local_id();
+
+        int offset[1 << RADIX_BITS];
+        const int radix_range = 1 << RADIX_BITS;
+        for (int i = 0; i < radix_range; i++)
+        {
+            offset[i] = 0;
+        }
+
+        for (int i = local_id; i < nSubgroupSums; i += local_size)
+        {
+            for (int j = 0; j < radix_range; j++)
+            {
+                int value                               = partialHists[i * radix_range + j];
+                int boundary                            = sub_group_scan_exclusive_add(value);
+                partialPrefixHists[i * radix_range + j] = offset[j] + boundary;
+                int partial_offset                      = sub_group_reduce_add(value);
+                offset[j] += partial_offset;
+            }
+        }
+
+        if (local_id == 0)
```

Comment:
Maybe it's possible to remove if () {} by using cycle like for (int j = local_id; j < radix_range; j += local_size). It should work because all local_id's have the same offsets. But it's not important

Reply:
removed this function from here and use it from services